torchgeo/datasets/sentinel.py

```diff
@@ -25,6 +25,221 @@ class Sentinel(RasterDataset):
+    # aaaaaaaa:   Start Date of Acquisition (YYYYMMDD)
+    # bbbbbb:     Start Time of Acquisition (HHMMSS)
+    # pp:         Polarization: Dual-pol (D) vs. Single-pol (S),
+    #                 primary polarization (H vs. V)
+    # o:          Orbit Type: Precise (P), Restituted (R), or Original Predicted (O)
+    # zz:         Terrain Correction Pixel Spacing
+    # u:          Software Package Used: GAMMA (G)
+    # d:          Gamma-0 (g) or Sigma-0 (s) Output
+    # e:          Power (p) or Decibel (d) or Amplitude (a) Output
+    # f:          Unmasked (u) or Water Masked (w)
+    # k:          Not Filtered (n) or Filtered (f)
+    # l:          Entire Area (e) or Clipped Area (c)
+    # m:          Dead Reckoning (d) or DEM Matching (m)
+    # ssss:       Product ID
+    filename_glob = "S1*{}.*"
+    filename_regex = r"""
+        ^S1(?P<mission>[AB])
+        _(?P<mode>SM|IW|EW|WV)
+        _(?P<date>\d{8}T\d{6})
+        _(?P<polarization>[DS][HV])
+        (?P<orbit>[PRO])
+        _RTC(?P<spacing>\d{2})
+        _(?P<package>G)
+        _(?P<output1>[gs])
+        (?P<output2>[pda])
+        (?P<mask>[uw])
+        (?P<filter>[nf])
+        (?P<area>[ec])
+        (?P<matching>[dm])
+        _(?P<product>[0-9A-Z]{4})
+        _(?P<band>[VH]{2})
+        \.
+    """
+    date_format = "%Y%m%dT%H%M%S"
+    all_bands = ["HH", "HV", "VV", "VH"]
+    separate_files = True
+
+    def __init__(
+        self,
+        root: str = "data",
+        crs: Optional[CRS] = None,
+        res: float = 10,
+        bands: Sequence[str] = ["VV", "VH"],
+        transforms: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+        cache: bool = True,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            crs: :term:`coordinate reference system (CRS)` to warp to
+                (defaults to the CRS of the first file found)
+            res: resolution of the dataset in units of CRS
+                (defaults to the resolution of the first file found)
+            bands: bands to return (defaults to ["VV", "VH"])
+            transforms: a function/transform that takes an input sample
+                and returns a transformed version
+            cache: if True, cache file handle to speed up repeated sampling
+
+        Raises:
+            AssertionError: if ``bands`` is invalid
+            FileNotFoundError: if no files are found in ``root``
+        """
+        assert len(bands) > 0, "'bands' cannot be an empty list"
+        assert len(bands) == len(set(bands)), "'bands' contains duplicate bands"
+
+        for band in bands:
+            assert band in self.all_bands, f"invalid band '{band}'"
+
+        # Sentinel-1 uses dual polarization, it can only transmit either
+        # horizontal or vertical at a single time
+        msg = """
+'bands' cannot contain both horizontal and vertical transmit at the same time.
```

Comment:
Why not? Is this limitation solely for plotting purposes? It is totally valid for multiple passes over the same area to be of different polarizations. It's also valid for consecutive captures from the satellite to be of different transmit/receive pairings. Additionally, the intensities for VH and HV are equal due to Helmholtz reciprocity!

Reply:
This limitation is inherent in our current `RasterDataset` implementation. Basically, every `GeoDataset` has an R-tree index containing a **single** filename for each spatiotemporal location. We populate this R-tree during `__init__` using a filename glob, which has to contain a **single** band name (can't be both HH and VV). Then, during `__getitem__`, we loop over **all** bands and swap the band in the filename to load the data. If all bands aren't present for every spatiotemporal location, then loading will fail. This was designed for MSI and HSI imagery without any thought towards (or knowledge of) SAR data.

There are a few options to deal with this:

### 1. Leave it up to the user

This is our current approach. The user can create two datasets and compute the union themselves.

Pro: simple
Con: users may have to do extra work

### 2. Compute the union for them

We could subclass `UnionDataset` instead of `RasterDataset` and create two `RasterDatasets` for the user. This is actually a very elegant solution and shouldn't be that difficult.

Pro: users don't have to do any extra work
Con: it would not be possible to change bands. You could create a dataset containing all 4 bands, but it would not be possible to create a dataset with only 2 bands

I assume horizontal transmit and vertical transmit are fundamentally different and some users may want to work with only one or the other but not both? This would not be possible with this idea. A workaround to support this would be to create 3 datasets:

* Sentinel1Horizontal: [HH, HV]: RasterDataset
* Sentinel1Vertical: [VV, VH]: RasterData
* Sentinel1: [HH, HV, VV, VH]: UnionDataset

### 3. Override `__init__` and `__getitem__` in Sentinel1

We could replace the default logic for populating the R-tree and loading the data to search for both HH and VV and to load [HH, HV] and [VV, VH].

Pro: would support 4 bands or 2 bands
Con: mostly duplicate code, more difficult to maintain, need to do this for every bi-pole SAR instrument

### 4. Modify RasterDataset to support optional bands

Instead of populating the R-tree with a single band for each spatiotemporal location, we could load all bands we can find. Then, at loading time, we only load the bands that were found.

Pro: supports both SAR as well as MSI/HSI where the user might not have downloaded all bands but forgot to change `bands`
Con: significantly slower since R-tree lookup will go from $\mathcal{O}(\log N)$ to $\mathcal{O}(\log CN)$ and memory will go from $\mathcal{O}(N)$ to $\mathcal{O}(CN)$ for all datasets. Even slower would be loading and grabbing the CRS and bounding box of every file from disk. Also, if some downloads have different numbers of bands, this will break when loading data. Easy fix for this would be to specify a subset of bands when instantiating the dataset, but the error message would not be clear and this may be difficult to figure out for an inexperienced user

Let me know how you feel about 1–4. I'm leaning towards 1 or 2. I'm hesitant to go with 3 because duplicate code == evil, and 4 would require a separate PR and longer discussion.

Reply:
I prefer 1 -- until we have someone who is actively using SAR imagery with TorchGeo let's not spend too much time making this class perfect